refactor!: abstract part of REPL into a base implementation

Snehil-Shah opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

- [x] divide `repl` into `repl/base` and `repl/cli` with 1:1 feature preservation
- [x] Decide approaches to solve problems with the refactor
- [x] Update documentation
- [x] Update workflows and make commands
- [x] Find & remove redundant modules
- [x] Final self-review


## Related Issues

> Does this pull request have any related issues?

No.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Snehil-Shah committed:
refactor!: move presentation into the `cli` namespace

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: passed
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
refactor!: abstract part of REPL into a base implementation

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: passed
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
refactor: move example

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Snehil-Shah commented:
I tried refactoring the REPL with the aim of preserving all functionality. It mostly works the same with some limitations.

- There are a couple of mostly "utility" modules that happen to be common between the base and cli implementation. I will investigate more ways to reduce this duplicacy.
- Completions for `settings` no longer work as the completer is implemented in the base implementation, and settings is implemented in the cli implementation. I see 2 approaches to curb this: find a way to allow inheritors of the base class to mutate the completer (by providing some configuration args, or maybe defining an entire completer class with methods allowing for configuration), or, let settings be implemented in the base class with the option given to inheritors to define their own set of settings. So cli can add settings for syntax highlighting etc, while jupyter can add for other stuff. Here too, we would have to make a dedicated settings class with methods and args that allows for defining settings.
- small limitation: earlier, in-REPL commands were highlighted differently than normal functions. That doesn't work now because some commands are defined in the base implementation, and some additional commands are defined in the cli implementation, and we aren't, atm, caching names of in-REPL commands. We can do that, although I don't think it's worth it for something so trivial. They are still highlighted, just the same as normal functions, and I think that's okay.
- the main problem: REPL documentation database. This is specific to commands like `help`, `example`, `info`, `tutorial`:
	- Firstly, commands like `tutorial` and `example` wouldn't work on jupyter with their current implementation, as they are tightly designed to work with CLI (tutorial uses presentation, and example takes control over filling the next code cells). This is a fixable problem, just define separate versions of the commands in the parent implementations (jupyter can simply print the examples).
	- Mainly, how do we manage the database? REPL-specific functions like `commands` also have their own help text. Some commands that are in the cli REPL might not be in the jupyter or any other inheritor's context. Hence having a single `help.json` that contains text for all commands from both `cli` and `jupyter` will be a mess. 
	Some solutions: help text resolution for REPL-specific commands can be execluded from our main database of aliases and their help text. Then, each inheritor can maintain their own json of help text for their commands, written by hand, no CI needed I suppose. The `help` command will then search in their local commands database, and then the main database. If we go this way, I guess we would have to define `help` and other commands in all inheritors themselves, as they have to be edited to look locally first and then query the base implementation using an internal `_docs` prototype method that returns the main database of aliases.. There's some code duplication there.

A lot of the trivial issues can be postponed for future PRs. Ideas are welcome.

Once this PR is ready for review, I will highlight and explain the notable changes for an easier review.

Snehil-Shah committed:
refactor: move commands to client implementation

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

Snehil-Shah commented:
@kgryte I have updated the refactoring. Couple of points:

- Completions don't work for settings and tutorials. Future work would be on making the completer extendable with plugins.

- REPL commands are highlighted the same as normal functions.

Rest everything works the same and all integration tests are passing on the cli REPL.

Should I also change the `make` commands and workflows with updated directories for `contributors.json` etc?

Snehil-Shah committed:
fix: remove `command` token from themes

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Snehil-Shah committed:
refactor: move base implementation to `ctor` package

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
fix: update redundant paths for make and cli

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: failed
---

Snehil-Shah committed:
fix: update import paths

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: failed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

Snehil-Shah committed:
build: update workflow paths

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

Snehil-Shah commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Snehil-Shah committed:
style: convert indents to 2 spaces

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
fix: revert redundant changes

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
docs: update package paths

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/repl/base/ctor/lib/main.js:
```diff
@@ -0,0 +1,591 @@
+* @returns {REPL} REPL instance
+*
+* @example
+* var REPL = require( '@stdlib/repl/base' );
+*
```
```suggestion
```

We don't require modules within the module itself.

kgryte commented on lib/node_modules/@stdlib/repl/base/ctor/lib/drain.js:
```diff
@@ -178,9 +156,7 @@ function drain( repl ) {
-			repl._ostream.write( 'Error: '+error.message+'\n' );
-			repl.emit( 'command', code.raw, false ); // command failed
-			displayPrompt( repl, false );
+			repl.emit( 'command', code.raw, false, error, false ); // command failed
```
What are the additional `error` and `false` arguments here? The `onCommand` listener, TMK, only has `cmd` and `success` arguments.

kgryte commented on lib/node_modules/@stdlib/repl/base/ctor/lib/main.js:
```diff
@@ -0,0 +1,591 @@
+	* @private
+	* @param {string} cmd - command
+	* @param {boolean} success - boolean indicating whether the command successfully executed
+	* @param {*} res - result
+	* @param {boolean} silent - boolean indicating whether the command output should be silenced
```
Okay. You are also passing along the result and a "silent" flag.

kgryte commented on lib/node_modules/@stdlib/repl/base/ctor/lib/main.js:
```diff
@@ -0,0 +1,591 @@
+setNonEnumerableReadOnly( REPL.prototype, '_isMultilineInput', isMultilineInput );
+
+/**
+* Processes a "raw" input command..
```
```suggestion
* Processes a "raw" input command.
```

kgryte commented on lib/node_modules/@stdlib/repl/base/ctor/lib/main.js:
```diff
@@ -0,0 +1,591 @@
+* @param {string} input - input command
+* @returns {boolean} boolean indicating whether the given input is a possible multi-line command
+*/
+setNonEnumerableReadOnly( REPL.prototype, '_isMultilineInput', isMultilineInput );
```
This is a new private method. Who is the intended caller?

kgryte commented on lib/node_modules/@stdlib/repl/base/ctor/lib/main.js:
```diff
@@ -0,0 +1,591 @@
+* @param {string} code - unevaluated command
+* @returns {(string|Error)} processed command or an error
+*/
+setNonEnumerableReadOnly( REPL.prototype, '_processCommand', processCommand );
```
This is a new private method. Who is the intended caller?

kgryte commented on lib/node_modules/@stdlib/repl/base/ctor/lib/main.js:
```diff
@@ -0,0 +1,591 @@
+* @param {string} cmd - command string
+* @returns {(Object|Error)} compiled command or an error
+*/
+setNonEnumerableReadOnly( REPL.prototype, '_compileCommand', compileCommand );
```
This is a new private method. Who is the intended caller?

kgryte replied:
I see this is being called by inheritors. If it is meant to be a public method, make it public. Don't force descendants to call private commands. A dev working on this should expect that anything prefixed with a `_` is fair game to change. By having downstream consumers call private methods, you break that contract.

That applies more generally. Simply because a descendant inherits from a parent, that does not mean a descendant should be privy to all the internals of a parent. If a parent classifies something as private, including internal data structures, those should be considered private even to descendants. If you need to provide info/data/methods to descendants, you need to create public APIs/contracts. Otherwise, you are creating tight coupling which is inherently brittle.

kgryte commented on lib/node_modules/@stdlib/repl/base/ctor/lib/main.js:
```diff
@@ -0,0 +1,591 @@
+	} else {
+		setAliasesGlobal( this._globalVars, this._aliases, context, ALIAS_OVERRIDES ); // eslint-disable-line max-len
+	}
+
```
You've dropped workspace snapshotting. Why?

Snehil-Shah replied:
I moved that to the inheritors as all workspace variables aren't loaded yet (commands etc).

kgryte commented on lib/node_modules/@stdlib/repl/base/ctor/lib/main.js:
```diff
@@ -0,0 +1,591 @@
+	} else {
+		setAliasesGlobal( this._globalVars, this._aliases, context, ALIAS_OVERRIDES ); // eslint-disable-line max-len
+	}
+
```
Previously, we sorted the list of globals. You've removed this. Why?

Snehil-Shah replied:
Moved it to the inheritors for the same reason as above.

kgryte commented on lib/node_modules/@stdlib/repl/base/ctor/lib/validate.js:
```diff
@@ -0,0 +1,119 @@
+* @private
+* @param {Object} opts - destination object
+* @param {Options} options - function options
+* @param {ReadableStream} [options.input=process.stdin] - input stream
```
You shouldn't be including default values in the docs here. Only in the main module entrypoint (or wherever defaults are actually set). This function does NOT supply the defaults.

kgryte commented on lib/node_modules/@stdlib/repl/base/package.json:
```diff
@@ -0,0 +1,62 @@
+    "terminal",
+    "shell",
+    "server",
+    "toplevel"
```
What is this keyword for?

kgryte commented on lib/node_modules/@stdlib/repl/cli/lib/commands/example.js:
```diff
@@ -162,7 +162,7 @@ function command( repl, cmds ) {
 
 					// If line is part of a multi-line input, wait for the next line...
-					if ( repl._multilineHandler.isMultilineInput( lines.join( '\n' ) ) ) {
+					if ( repl._isMultilineInput( lines.join( '\n' ) ) ) {
```
You're calling a private method of the private. Don't do this. Make it a public method.

kgryte commented on lib/node_modules/@stdlib/repl/cli/lib/eager_evaluator.js:
```diff
@@ -301,13 +299,13 @@ setNonEnumerableReadOnly( EagerEvaluator.prototype, 'onKeypress', function onKey
 	}
 	debug( 'Processing command...' );
-	tmp = processCommand( code );
+	tmp = this._repl._processCommand( code );
```
Don't call private methods on the parent. Make a public method.

kgryte commented on lib/node_modules/@stdlib/repl/cli/lib/eager_evaluator.js:
```diff
@@ -301,13 +299,13 @@ setNonEnumerableReadOnly( EagerEvaluator.prototype, 'onKeypress', function onKey
 	}
 	debug( 'Compiling command...' );
-	executable = compileCommand( tmp );
+	executable = this._repl._compileCommand( tmp );
```
Don't call private methods on the parent. Make a public method.

kgryte commented on lib/node_modules/@stdlib/repl/cli/lib/syntax_highlighter.js:
```diff
@@ -329,7 +328,7 @@ setNonEnumerableReadOnly( SyntaxHighlighter.prototype, 'onKeypress', function on
 		debug( 'Line change detected. Tokenizing line: %s', this._rli.line );
 		this._line = this._rli.line; // update line buffer
-		tokens = tokenizer( this._line, this._repl._context );
+		tokens = this._repl._tokenizer( this._line, this._repl._context );
```
Don't call private methods on the parent. Make a public method.

kgryte commented on lib/node_modules/@stdlib/repl/cli/lib/multiline_handler.js:
```diff
@@ -739,7 +631,7 @@ setNonEnumerableReadOnly( MultilineHandler.prototype, 'beforeKeypress', function
 
 		// If we are in paste mode or the command is incomplete, trigger multi-line mode...
-		if ( !this._multiline.pasteMode && !this.isMultilineInput( cmd.join( '\n' ) ) ) {
+		if ( !this._multiline.pasteMode && !this._repl._isMultilineInput( cmd.join( '\n' ) ) ) {
```
Don't call private methods on the parent. Make a public method.

kgryte commented on lib/node_modules/@stdlib/repl/cli/lib/main.js:
```diff
@@ -481,11 +417,9 @@ function REPL( options ) {
-			return;
-		}
+		// Emit an interrupt:
+		self.emit( 'interrupt' );
```
Why is this necessary? Should add a command providing an explanation.

kgryte commented on lib/node_modules/@stdlib/repl/cli/lib/main.js:
```diff
@@ -523,13 +454,28 @@ function REPL( options ) {
+		if ( success === false ) {
+			self._ostream.write( 'Error: '+res.message+'\n' );
+		} else if ( !silent && res !== void 0 ) {
+			pre = replace( self._outputPrompt, '%d', (self._count).toString() );
```
```suggestion
			pre = replace( self._outputPrompt, '%d', self._count.toString() );
```

kgryte replied:
`_count` is another magical private variable. This makes it very difficult to reason about class behavior. Yes, `@stdlib/repl/presentation` peers into `@stdlib/repl/lib`, but that doesn't mean it was a good thing or the right decision.

kgryte commented on lib/node_modules/@stdlib/repl/cli/lib/main.js:
```diff
@@ -729,17 +675,8 @@ setNonEnumerableReadOnly( REPL.prototype, 'createContext', function createContex
-	} else {
-		setAliasesGlobal( this._globalVars, this._aliases, context, ALIAS_OVERRIDES ); // eslint-disable-line max-len
-	}
+	// TODO: Call parent method
```
Why is this TODO here?

kgryte commented:
In general, we need a cleaner separation between the base class. We don't want child classes examining the private contents of the private and then doing various undesired things. In short, if a child needs something from the parent, the parent needs to provide a public API/contract. E.g., if a child needs to know the command count, well, the parent needs to provide a property accessor for returning the command count. No accessing `_count`. This is magic and makes it very difficult for future devs to know when and where particular properties are defined.

Snehil-Shah committed:
fix: apply review comments

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
docs: update package keywords

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
refactor!: move `repl/cli` to `repl`

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
refactor!: workspaces

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
docs: fix examples

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

Snehil-Shah committed:
refactor: separate concern of streams

The base ctor needs the stream to feed to the Console object. The CLI implementation needs it to paint the entire UI. They serve different purposes in both parties, and hence exposing APIs from the base implementation that allows access to the underlying ostream just doesn't make sense. It's duplication, but cleanly separates concerns

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
refactor: clean up context creation

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
fix: heal examples

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
docs: document `setCommands`

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

Snehil-Shah committed:
refactor: evalin command

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Snehil-Shah committed:
refactor: move `_cmd` and run logic

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

Snehil-Shah committed:
refactor: move `_history` and create accessor for count

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

Snehil-Shah committed:
refactor: abstract alias resolution

Signed-off-by: Snehil Shah <snehilshah.989@gmail.com>

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---
